bucket.go

```diff
@@ -9,86 +9,127 @@ import (
 	"github.com/libp2p/go-libp2p-core/peer"
 )
 
-// Bucket holds a list of peers.
-type Bucket struct {
+// PeerState is the state of the peer as seen by the Routing Table.
+type PeerState int
+
+const (
+	// PeerStateActive indicates that we know the peer is active/alive.
+	PeerStateActive PeerState = iota
+	// PeerStateMissing indicates that we do not know the state of the peer.
+	PeerStateMissing
+)
+
+// PeerInfo holds all related information for a peer in the K-Bucket.
+type PeerInfo struct {
+	Id    peer.ID
+	State PeerState
+}
+
+// bucket holds a list of peers.
+type bucket struct {
 	lk   sync.RWMutex
 	list *list.List
 }
 
-func newBucket() *Bucket {
-	b := new(Bucket)
+func newBucket() *bucket {
+	b := new(bucket)
 	b.list = list.New()
 	return b
 }
 
-func (b *Bucket) Peers() []peer.ID {
+// returns all peers in the bucket
+func (b *bucket) peers() []PeerInfo {
+	b.lk.RLock()
+	defer b.lk.RUnlock()
+	var ps []PeerInfo
+	for e := b.list.Front(); e != nil; e = e.Next() {
+		p := e.Value.(PeerInfo)
+		ps = append(ps, p)
+	}
+	return ps
+}
+
+// return the Ids of all the peers in the bucket.
+func (b *bucket) peerIds() []peer.ID {
 	b.lk.RLock()
 	defer b.lk.RUnlock()
 	ps := make([]peer.ID, 0, b.list.Len())
 	for e := b.list.Front(); e != nil; e = e.Next() {
-		id := e.Value.(peer.ID)
-		ps = append(ps, id)
+		p := e.Value.(PeerInfo)
+		ps = append(ps, p.Id)
 	}
 	return ps
 }
 
-func (b *Bucket) Has(id peer.ID) bool {
+// returns the peer with the given Id and true if peer exists
+// returns false if the peerId does not exist
+func (b *bucket) getPeer(p peer.ID) (PeerInfo, bool) {
 	b.lk.RLock()
 	defer b.lk.RUnlock()
 	for e := b.list.Front(); e != nil; e = e.Next() {
-		if e.Value.(peer.ID) == id {
+		if e.Value.(PeerInfo).Id == p {
+			return e.Value.(PeerInfo), true
+		}
+	}
+	return PeerInfo{}, false
+}
+
+// replaces the peer based on the Id.
+// returns true if the replace was successful, false otherwise.
+func (b *bucket) replace(p PeerInfo) bool {
```

Comment:
Why not just use a pointer? That is, have an unlocked variant of getPeer that returns a pointer to the peer info that we then update in-place?

Reply:
Note: we might as well store peer infos by pointer anyways as the `interface{}` indirection will force us to put it on the heap regardless.

Reply:
Ah, are we intentionally moving the peer to the back?

Reply:
@Stebalien I agree. Using a pointer here makes sense. While I was intentionally moving the peer to the back, I don't think there's any point in doing it. Take a look at https://github.com/libp2p/go-libp2p-kbucket/pull/50#discussion_r379993597.

I'm  making this change.

Reply:
I've made the change.